blockscout-ens/bens-logic/src/protocols/protocoler.rs

```diff
@@ -283,16 +283,94 @@ impl Protocoler {
         Ok(protocols)
     }
 
+    pub fn fetch_domain_options(
+        &self,
+        name_with_tld: &str,
+        network_id: i64,
+        maybe_filter: Option<NonEmpty<String>>,
+    ) -> Result<Vec<DomainNameOnProtocol>, ProtocolError> {
+        let tld = self.extract_tld(name_with_tld)?;
+
+        let protocols = self.protocols_of_network_for_tld(network_id, tld, maybe_filter)?;
+        let domain_names: Vec<_> = protocols
+            .into_iter()
+            .filter_map(|p| DomainNameOnProtocol::from_str(name_with_tld, p).ok())
+            .collect();
+
+        if domain_names.is_empty() {
+            return Err(ProtocolError::InvalidName {
+                name: name_with_tld.to_string(),
+                reason: "No matching protocols for given TLD".to_string(),
+            });
+        }
+
+        Ok(domain_names)
+    }
+
+    pub fn lookup_names_options_in_network(
+        &self,
+        name: &str,
+        network_id: i64,
+        maybe_filter: Option<NonEmpty<String>>,
+    ) -> Result<Vec<DomainNameOnProtocol>, ProtocolError> {
+        let protocols = self
+            .networks
+            .get(&network_id)
+            .ok_or_else(|| ProtocolError::NetworkNotFound(network_id))?
+            .use_protocols
+            .iter()
+            .filter_map(|proto_name| self.protocols.get(proto_name))
+            .collect::<Vec<_>>();
+
+        for proto in &protocols {
+            let empty_hash = match &proto.info.protocol_specific {
+                ProtocolSpecific::EnsLike(cfg) => cfg.empty_label_hash,
+                _ => None,
+            };
+
+            if let Ok(domain) = DomainName::new(name, empty_hash) {
+                if domain.level_gt_tld() {
+                    let fetched =
+                        self.fetch_domain_options(&domain.name, network_id, maybe_filter)?;
+                    return Ok(fetched.into_iter().collect());
+                }
+            }
+        }
```

Comment:
why do you need `fetched.into_iter().collect()` ?

also you can simplify this
```suggestion
            if let Ok(domain) = DomainName::new(name, empty_hash).map(|name| name.level_gt_tld()).unwrap_or_default() {
                        self.fetch_domain_options(&domain.name, network_id, maybe_filter)
                }
        }
```